### DISCUSSION TITLE
SESSION_STORE vs SESSION_DRIVER: same thing?

### DISCUSSION THREAD

**Discussion by bert-w** [laravel/framework#41913] [Q&A]
Hi

There seem to be 2 configurations named `SESSION_DRIVER` and `SESSION_STORE` in `config/session.php` as seen here:
 - https://github.com/laravel/laravel/blob/9.x/config/session.php#L103
 - https://github.com/laravel/laravel/blob/9.x/config/session.php#L21

However, they seem to do the exact same thing. When I look at the SessionManager, it just looks like `config('session.store')` takes precedence over the other value:

https://github.com/laravel/framework/blob/78cb588f9f391016632b3f96755c851a712a442f/src/Illuminate/Session/SessionManager.php#L163-L177

Anyone know whats up, or what the use case is for these 2 separate variables?

**Comment by veelasky** [comment]
the purpose of it is to be able to configure multiple session store with the same driver. say you have `redis` as the driver, you can configure multiple session store for different `redis` configuration

**Reply by bert-w** [reply]
I dont see how that works, there sadly also isn't any documentation for it.

say I put `config/session.php::'session.driver'=>'redis'` and `config/session.php::'session.store' => null`. The SessionManager will run the setup for Redis, and since `session.store` is null, it will just use the driver name and use `'redis'`.

Now, this `'redis'` refers to a key in `config/cache.php::'stores'`, which states:
```php
'redis' => [
    'driver' => 'redis',
    'connection' => 'cache', // refers to a key in `config/database.php::'redis'`
],
```

This configuration states that the driver is `'redis'` and the connection is `'cache'`. The silly thing is, those are definable using `config/session.php::'driver' => env('SESSION_DRIVER')` and `config/session.php::'connection' => env('SESSION_CONNECTION')`. So the `SESSION_CONNECTION` seems to be some sort of override for the configuration in `config/cache.php::'stores'`.

Now when I would change `session.store = 'mycustomredis'`, I would have to define a new key in `config/cache.php::'stores'` named `'mycustomredis'`, which contains the same code as the snippet above, only with a different `connection` for instance (which then refers to a key in `config/database.php::'redis'`). But I could just as well change the `SESSION_CONNECTION` variable right?

**Reply by decadence** [reply]
@bert-w Did you find purpose of the SESSION_STORE? I don't get why it's needed too.

**Comment by pakogn** [comment]
Hello,

a bit late but I noticed this today.
It appears that Laravel uses a cache in sessions, so the SESSION_STORE is this cache store thing. If you go to cache config file you'll see these stores. So is to define an specific cache store of the cache config file. This is independent from the SESSION_DRIVER and SESSION_CONNECTION (database config file connection)

**Comment by dmitry-ivanov** [comment]
The SESSION_DRIVER defines the mechanism that will be used to store sessions _(f.e., SESSION_DRIVER=redis)_.

The SESSION_CONNECTION defines the specific connection _(for the "database" or "redis" drivers)_.
It is defined in the `config/database.php` _(see "connections" and "redis" array keys)_.

The SESSION_STORE is an _**alternative**_ to the SESSION_CONNECTION.
It defines the specific cache store for the cache-driven drivers only _("apc", "dynamodb", "memcached", "redis")_.
It is defined in the `config/cache.php` _(see "stores" array key)_.

By using a store, you can define _NOT ONLY_ the connection but also additional parameters _(e.g., a prefix for all keys stored)_.

## SESSION_CONNECTION vs SESSION_STORE? 🤔

Use SESSION_CONNECTION if you need to control only the host/port/database:
```php
'sessions' => [
    'url' => env('REDIS_URL'),
    'host' => env('REDIS_HOST', '127.0.0.1'),
    'username' => env('REDIS_USERNAME'),
    'password' => env('REDIS_PASSWORD'),
    'port' => env('REDIS_PORT', '6379'),
    'database' => env('REDIS_SESSION_DB', '2'),
],
```

Use SESSION_STORE if the separation that SESSION_CONNECTION gives is NOT ENOUGH for you, and you need to specify additional params _(prefix, etc.)_

_(For example, you'd want to use SESSION_STORE when your staging and production use the same Redis connection, and you want to force different prefixes for them)_
